lib/src/components/AbstractDialog/AbstractDialog.tsx

```diff
@@ -0,0 +1,213 @@
+        // HTML <dialog> elements are supposed to remember what element was focused before they were opened,
+        // and restore focus to that element when they are closed. No browsers appear to implement this currently,
+        // so we do it ourselves
+        previouslyFocusedEl = document.activeElement;
+
+    setDialogRefState(el);
+
+    /*
+     * This will cause the document to become "blocked by the modal dialog"
+     * (https://html.spec.whatwg.org/multipage/interaction.html#blocked-by-a-modal-dialog)
+     * meaning that only the modal and its contents are interactable/focusable.
+     *
+     * Note: not supported in safari, which is why the conditional is here. Once safari gets
+     * it together, we should be able to take advantage of the "top layer" functionality built into
+     * the browser around modals to simplify the dialog styling around z-index handling
+     */
+    if (hasNativeModalSupport) {
+      // https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1029#issuecomment-968299542
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (el as any).showModal();
+    }
+    else {
+      // without native support we don't trap focus in the dialog, but we can at least start it off there
+      el.focus();
+    }
+
+    return () => {
+      if (previouslyFocusedEl && previouslyFocusedEl instanceof HTMLElement) {
+        // The useEffect cleanup executes while the dialog is still present (in React 16 at least). While the dialog
+        // still exists, the document is still "blocked by the modal dialog" so trying to focus elements outside of
+        // it won't work. So we have to wait until the next cycle of the event loop when it's gone
+        Promise.resolve().then(() => previouslyFocusedEl.focus());
+      }
+    };
+  }, []);
+
+  // listen to the native HTMLDialogElement cancel event which supporting browsers fire when the dialog is closed
+  // via ESC
+  useEffect(function() {
+    const dialog = dialogRef.current;
+
+    if (dialog && hasNativeModalSupport) {
+      /* eslint-disable @typescript-eslint/no-non-null-assertion */
+      dialog.addEventListener('cancel', onCancel!);
+      return () => { dialog.removeEventListener('cancel', onCancel!); };
+      /* eslint-enable @typescript-eslint/no-non-null-assertion */
+    }
+    else {
+      return undefined;
+    }
+  }, [onCancel]);
+
+  if (cancelOnClickOutside) {
+    const clickOutsideTargetElement = cancelOnClickOutsideTargetClassName ?
+      dialogRef.current?.getElementsByClassName(cancelOnClickOutsideTargetClassName)[0] :
+      dialogRef.current;
+
+    useEffect(() => {
+      const listener = (event: MouseEvent) => {
+        if (!clickOutsideTargetElement || clickOutsideTargetElement.contains(event.target as Node)) {
+          return;
+        }
+
+        if (onCancel) {
+          onCancel(createCancelEvent());
+        }
+      };
+
+      document.addEventListener('click', listener);
```

Comment:
You might consider doing what modal does, having the `<dialog>` actually be a "mask" element (which might be totally transparent in this case) which covers the entire viewport.  Then you could attach a regular react `onClick` prop to that element rather than having to use a native event listener

Reply:
> You might consider doing what modal does, having the `<dialog>` actually be a "mask" element (which might be totally transparent in this case) which covers the entire viewport. Then you could attach a regular react `onClick` prop to that element rather than having to use a native event listener

Still in progress but we've done that here: e9e1ab56141269a61c3331076136ed998c521595